File: BOB/Hitter/RobotBridge4_refactor/deploy/mocap_bridge/tests/test_transformation_t_v2.py
import subprocess
from pathlib import Path

import pytest

from unitree_sdk2.lcm_types.transformation_t import transformation_t


def test_new_messages_default_track_id_to_zero():
    """A missing v2 track field would make the generated binding unusable."""
    assert transformation_t().track_id == 0


def test_python_round_trip_preserves_track_id():
    """Dropping or misordering track_id on the v2 wire must be observable."""
    msg = transformation_t()
    msg.name = "ball"
    msg.vicon_frame_number = 41
    msg.vicon_time_s = 0.125
    msg.publish_time_us = 1_700_000_000_000_000
    msg.track_id = 9001
    msg.valid, msg.occluded = 1, 0
    msg.pos_vicon = [0.7, -0.2, 1.0]
    msg.quat_vicon = [0.0, 0.0, 0.0, 1.0]

    decoded = transformation_t.decode(msg.encode())

    assert (
        decoded.name,
        decoded.vicon_frame_number,
        decoded.vicon_time_s,
        decoded.publish_time_us,
        decoded.track_id,
        decoded.valid,
        decoded.occluded,
        decoded.pos_vicon,
        decoded.quat_vicon,
    ) == (
        "ball",
        41,
        0.125,
        1_700_000_000_000_000,
        9001,
        1,
        0,
        (0.7, -0.2, 1.0),
        (0.0, 0.0, 0.0, 1.0),
    )


def test_v1_payload_is_rejected_by_the_v2_fingerprint():
    """A v1 message must fail on its known fingerprint, not a truncated body."""
    v1_payload_bytes = bytes.fromhex("71f936e3b20f1df5") + (b"\0" * 96)

    assert transformation_t._get_packed_fingerprint() != v1_payload_bytes[:8]
    with pytest.raises(ValueError, match="Decode error"):
        transformation_t.decode(v1_payload_bytes)


def test_python_decodes_fixed_cpp_v2_wire_fixture():
    """The C++ producer's fixed v2 payload must decode in the Python binding."""
    binary = Path(__file__).parents[1] / ".build-v2" / "test_transformation_t_v2"
    if not binary.is_file():
        pytest.skip("C++ v2 fixture is built by build_v2_mocap.sh")

    completed = subprocess.run([str(binary), "--emit-hex"], check=True, capture_output=True, text=True)
    decoded = transformation_t.decode(bytes.fromhex(completed.stdout.strip()))

    assert decoded.track_id == 9001

File: BOB/Hitter/RobotBridge4/deploy/mocap_bridge/tests/test_transformation_t_v2.cpp
#include "unitree_sdk2/lcm_types/transformation_t.hpp"

#include <iomanip>
#include <iostream>
#include <string>
#include <vector>

int main(int argc, char** argv) {
  lcm_types::transformation_t msg{};
  msg.name = "ball";
  msg.vicon_frame_number = 41;
  msg.vicon_time_s = 0.125;
  msg.publish_time_us = 1700000000000000LL;
  msg.track_id = 9001;
  msg.valid = 1;
  msg.occluded = 0;
  msg.pos_vicon[0] = 0.7;
  msg.pos_vicon[1] = -0.2;
  msg.pos_vicon[2] = 1.0;
  msg.quat_vicon[3] = 1.0;

  std::vector<unsigned char> wire(msg.getEncodedSize());
  if (msg.encode(wire.data(), 0, wire.size()) < 0) return 2;
  if (argc == 2 && std::string(argv[1]) == "--emit-hex") {
    for (unsigned char byte : wire) {
      std::cout << std::hex << std::setw(2) << std::setfill('0') << int(byte);
    }
    std::cout << "\n";
  }
  return msg.track_id == 9001 ? 0 : 3;
}
